Consolidate to one `schedule` param

dstandish opened the pull request:
Consolidate the three scheduling params to one param `schedule`

dstandish committed:
Rename DAG `schedule_on` parameter to `schedule`

The intention is that `schedule` becomes the single DAG scheduling param, and that we will deprecate `schedule_interval` and `timetable`.

dstandish committed:
add deprecation warning for old params

dstandish committed:
harvest schedule interval args and timetables from `schedule`

dstandish committed:
fixup! harvest schedule interval args and timetables from `schedule`

dstandish committed:
fix runtime checkable import

dstandish committed:
rename most usages of schedule_interval= to schedule=

dstandish committed:
stacklevel

dstandish committed:
fix param handling

dstandish committed:
update some docs

dstandish committed:
update docs

dstandish committed:
fix tests

dstandish committed:
fix test dag endpoint

dstandish committed:
fix dag run endpoint test

dstandish committed:
fix param namee

eladkal commented on docs/apache-airflow/templates-ref.rst:
```diff
@@ -89,16 +89,11 @@ existing code to use other variables instead.
 =====================================   ====================================
 ``{{ execution_date }}``                the execution date (logical date), same as ``dag_run.logical_date``
-``{{ next_execution_date }}``           the next execution date (if available) (`pendulum.DateTime`_)
-                                        if ``{{ execution_date }}`` is ``2018-01-01 00:00:00`` and
-                                        ``schedule_interval`` is ``@weekly``, ``{{ next_execution_date }}``
-                                        will be ``2018-01-08 00:00:00``
```
I don't think the intent was to remove the example?

dstandish replied:
it was intentional.  firstly, these are all *deprecated*, so folks reading the docs should not be using them anyway.  secondly, by referencing "logical date"  in the definition, we defer to the example in the section of non-deprecated items.

note there is also no example for the also-deprecated value `execution_date`

the example and explanation were confusing anyway so this to me  seem better

dstandish committed:
fix type

dstandish committed:
fix tests

dstandish committed:
fix tests

dstandish committed:
provider examples should stick with schedule_interval for now

dstandish committed:
remove schedule_interval param from example dags

dstandish committed:
fixup! remove schedule_interval param from example dags

dstandish committed:
fix test

jedcunningham commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -82,7 +81,7 @@ Data Interval
 Each DAG run in Airflow has an assigned "data interval" that represents the time
 range it operates in. For a DAG scheduled with ``@daily``, for example, each of
-its data interval would start each day at midnight (00:00) and end at midnight
+ts data interval would start each day at midnight (00:00) and end at midnight
```
```suggestion
its data interval would start each day at midnight (00:00) and end at midnight
```

jedcunningham commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -117,8 +116,8 @@ DAG run fails.
 
-An Airflow DAG with a ``start_date``, possibly an ``end_date``, and a ``schedule_interval`` defines a
-series of intervals which the scheduler turns into individual DAG Runs and executes. The scheduler, by default, will
+An Airflow DAG defined with a ``start_date``, possibly an ``end_date``, and a either a cron expression or timetable, defines a series of intervals which the scheduler turns into individual DAG Runs and executes.
```
```suggestion
An Airflow DAG defined with a ``start_date``, possibly an ``end_date``, and either a cron expression or timetable, defines a series of intervals which the scheduler turns into individual DAG Runs and executes.
```

I think the sentence still need a little love.

jedcunningham commented on docs/apache-airflow/faq.rst:
```diff
@@ -41,13 +41,13 @@ There are very many reasons why your task might not be getting scheduled. Here a
 
-- Is your ``start_date`` set properly? The Airflow scheduler triggers the
-  task soon after the ``start_date + schedule_interval`` is passed.
+- Is your ``start_date`` set properly? The Airflow scheduler won't trigger the task until the
```
```suggestion
- Is your ``start_date`` set properly? The Airflow scheduler won't trigger the task until
```

jedcunningham commented on docs/apache-airflow/faq.rst:
```diff
@@ -41,13 +41,13 @@ There are very many reasons why your task might not be getting scheduled. Here a
-- Is your ``start_date`` set properly? The Airflow scheduler triggers the
-  task soon after the ``start_date + schedule_interval`` is passed.
+- Is your ``start_date`` set properly? The Airflow scheduler won't trigger the task until the
+  after the first schedule interval following the start date has passed.
```
Might be worth adding a simple example here, since the "start_date+schedule_interval" doesn't work as cleanly any longer.

dstandish replied:
reworded to clarify context

jedcunningham commented on newsfragments/25410.significant.rst:
```diff
@@ -0,0 +1,3 @@
+Deprecation of ``schedule_interval`` and ``timetable`` params
+
+We add new DAG parameter ``schedule`` that can accept a cron expression, timedelta object, *timetable* object, or list of dataset objects. Params ``schedule_interval`` and ``timetable`` are  deprecated.
```
Might be worth adding a code example, "this becomes this" type thing.

dstandish replied:
done

jedcunningham commented on scripts/in_container/verify_providers.py:
```diff
@@ -200,6 +200,7 @@ class ProviderPackageDetails(NamedTuple):
     "distutils Version classes are deprecated. Use packaging.version instead.",
     "the imp module is deprecated in favour of importlib; "
     "see the module's documentation for alternative uses",
+    "see the module's documentation for alternative uses",
```
```suggestion
```

dstandish committed:
Update scripts/in_container/verify_providers.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update docs/apache-airflow/faq.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update docs/apache-airflow/dag-run.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update docs/apache-airflow/dag-run.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
add example

dstandish committed:
simplify example

jedcunningham commented on newsfragments/25410.significant.rst:
```diff
@@ -0,0 +1,43 @@
+    with DAG(
+        dag_id='my_example',
+        start_date=datetime(2021, 1, 1),
+        timetable=EventsTimetable(event_dates=[pendulum.datetime(2022, 4, 5)]),
```
```suggestion
        schedule=EventsTimetable(event_dates=[pendulum.datetime(2022, 4, 5)]),
```

dstandish committed:
add test

dstandish committed:
Update newsfragments/25410.significant.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

uranusjr committed:
Simplify and fix test

uranusjr commented on tests/test_utils/perf/dags/elastic_dag.py:
```diff
@@ -175,13 +175,13 @@ class DagShape(Enum):
                     f"TASKS_COUNT=${TASKS_COUNT}",
                     f"START_DATE=${START_DATE_ENV}",
-                    f"SCHEDULE_INTERVAL=${SCHEDULE_INTERVAL_ENV}",
+                    f"schedule=${SCHEDULE_INTERVAL_ENV}",
```
Not quite sure what this one is for, it’s not actually DAG arguments.

dstandish replied:
i guess it's to generate a dag_id. i'll make it uppercase.

dstandish committed:
uppercase

dstandish committed:
fix log testing

uranusjr committed:
Unify comment format

uranusjr commented on newsfragments/25410.significant.rst:
```diff
@@ -0,0 +1,43 @@
+
+We add new DAG parameter ``schedule`` that can accept a cron expression, timedelta object, *timetable* object, or list of dataset objects. Params ``schedule_interval`` and ``timetable`` are  deprecated.
+
+So if you previously used the ``@daily`` cron preset, your dag may have looked like this:
```
```suggestion
If you previously used the ``@daily`` cron preset, your DAG may have looked like this:
```

uranusjr commented on newsfragments/25410.significant.rst:
```diff
@@ -0,0 +1,43 @@
+    ) as dag:
+        ...
+
+Going forward, instead you should use the ``schedule`` param instead:
```
```suggestion
Going forward, you should use the ``schedule`` argument instead:
```

uranusjr commented on newsfragments/25410.significant.rst:
```diff
@@ -0,0 +1,43 @@
+    ) as dag:
+        ...
+
+And the same is true if you used a custom timetable.  Previously you would have used the ``timetable`` param:
```
```suggestion
The same is true if you used a custom timetable.  Previously you would have used the ``timetable`` argument:
```

uranusjr commented on newsfragments/25410.significant.rst:
```diff
@@ -0,0 +1,43 @@
+    ) as dag:
+        ...
+
+Now you should use the ``schedule`` param:
```
```suggestion
Now you should use the ``schedule`` argument:
```

uranusjr commented on newsfragments/25410.significant.rst:
```diff
@@ -0,0 +1,43 @@
+Deprecation of ``schedule_interval`` and ``timetable`` params
```
```suggestion
Deprecation of ``schedule_interval`` and ``timetable`` arguments
```

uranusjr commented on newsfragments/25410.significant.rst:
```diff
@@ -0,0 +1,43 @@
+Deprecation of ``schedule_interval`` and ``timetable`` params
+
+We add new DAG parameter ``schedule`` that can accept a cron expression, timedelta object, *timetable* object, or list of dataset objects. Params ``schedule_interval`` and ``timetable`` are  deprecated.
```
```suggestion
We added new DAG argument ``schedule`` that can accept a cron expression, timedelta object, *timetable* object, or list of dataset objects. Arguments ``schedule_interval`` and ``timetable`` are deprecated.
```

uranusjr commented on docs/apache-airflow/concepts/datasets.rst:
```diff
@@ -37,11 +37,11 @@ Then reference the dataset as a task outlet:
     :end-before: [END task_outlet]
 
-Finally, define a DAG and reference this dataset in the DAG's ``schedule_on`` parameter:
+Finally, define a DAG and reference this dataset in the DAG's ``schedule`` parameter:
```
```suggestion
Finally, define a DAG and reference this dataset in the DAG's ``schedule`` argument:
```

uranusjr commented on docs/apache-airflow/concepts/datasets.rst:
```diff
@@ -37,11 +37,11 @@ Then reference the dataset as a task outlet:
     :end-before: [END dag_dep]
 
-You can reference multiple datasets in the DAG's ``schedule_on`` param.  Once there has been an update to all of the upstream datasets, the DAG will be triggered.  This means that the DAG will run as frequently as its least-frequently-updated dataset.
+You can reference multiple datasets in the DAG's ``schedule`` param.  Once there has been an update to all of the upstream datasets, the DAG will be triggered.  This means that the DAG will run as frequently as its least-frequently-updated dataset.
```
```suggestion
You can reference multiple datasets in the DAG's ``schedule`` argument.  Once there has been an update to all of the upstream datasets, the DAG will be triggered.  This means that the DAG will run as frequently as its least-frequently-updated dataset.
```

uranusjr commented on docs/apache-airflow/concepts/timetable.rst:
```diff
@@ -19,16 +19,18 @@
-as described in :doc:`DAG Run </dag-run>`, or by passing a ``timetable`` argument
-directly. The timetable also dictates the data interval and the logical time of each
-run created for the DAG.
+For dags with time-based schedules (as opposed to event-driven), the scheduling
```
```suggestion
For DAGs with time-based schedules (as opposed to event-driven), the scheduling
```

uranusjr commented on docs/apache-airflow/concepts/timetable.rst:
```diff
@@ -19,16 +19,18 @@
-directly. The timetable also dictates the data interval and the logical time of each
-run created for the DAG.
+For dags with time-based schedules (as opposed to event-driven), the scheduling
+decisions are driven by its internal "timetable".  The timetable class also
```
```suggestion
decisions are driven by its internal "timetable".  The timetable also
```

uranusjr commented on docs/apache-airflow/concepts/timetable.rst:
```diff
@@ -19,16 +19,18 @@
-respectively), however, there are situations where they cannot properly express
-the schedule. Some examples are:
+Even for dags whose scheduling is defined by a cron expression or ``timedelta`` object,
+internally these expressions converted to a timetable.
```
```suggestion
DAGs scheduled with a cron expression or ``timedelta`` object are
internally converted to always use a timetable.
```

uranusjr commented on docs/apache-airflow/concepts/timetable.rst:
```diff
@@ -19,16 +19,18 @@
+
+If a cron expression or ``timedelta`` is sufficient for your use case, you don't need
+to worry about timetables.  But for more complicated scheduling requirements,
+you may create your own timetable class and pass that to the DAG ``schedule`` param.
```
```suggestion
If a cron expression or ``timedelta`` is sufficient for your use case, you don't need
to worry about timetables.  But for more complicated scheduling requirements,
you may create your own timetable class and pass that to the DAG's ``schedule`` argument instead.
```

uranusjr commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -43,10 +43,9 @@ There are two possible terminal states for the DAG Run:
-a ``str``, a ``datetime.timedelta`` object, or one of the following cron "presets".
+You may set your DAG to run on a simple schedule by setting its ``schedule`` param with either a
+`cron expression <https://en.wikipedia.org/wiki/Cron#CRON_expression>`_, or a ``datetime.timedelta`` object,
+or one of the following cron "presets". For more elaborate scheduling requirements, you can implement a :doc:`custom timetable </concepts/timetable>`
```
```suggestion
You may set your DAG to run on a simple schedule by setting its ``schedule`` argument to either a
`cron expression <https://en.wikipedia.org/wiki/Cron#CRON_expression>`_, a ``datetime.timedelta`` object,
or one of the following cron "presets". For more elaborate scheduling requirements, you can implement a :doc:`custom timetable </concepts/timetable>`
```

uranusjr commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -117,8 +116,8 @@ DAG run fails.
 
-An Airflow DAG with a ``start_date``, possibly an ``end_date``, and a ``schedule_interval`` defines a
-series of intervals which the scheduler turns into individual DAG Runs and executes. The scheduler, by default, will
+An Airflow DAG defined with a ``start_date``, possibly an ``end_date``, and either a cron expression or timetable, defines a series of intervals which the scheduler turns into individual DAG Runs and executes.
```
```suggestion
An Airflow DAG defined with a ``start_date``, possibly an ``end_date``, and a non-dataset schedule, defines a series of intervals which the scheduler turns into individual DAG runs and executes.
```

uranusjr commented on docs/apache-airflow/faq.rst:
```diff
@@ -41,13 +41,13 @@ There are very many reasons why your task might not be getting scheduled. Here a
 
-- Is your ``start_date`` set properly? The Airflow scheduler triggers the
-  task soon after the ``start_date + schedule_interval`` is passed.
+- Is your ``start_date`` set properly? For time-based dags, the task won't be triggered until the
```
```suggestion
- Is your ``start_date`` set properly? For time-based DAGs, the task won't be triggered until the
```

uranusjr commented on docs/apache-airflow/faq.rst:
```diff
@@ -41,13 +41,13 @@ There are very many reasons why your task might not be getting scheduled. Here a
 
-- Is your ``schedule_interval`` set properly? The default ``schedule_interval``
-  is one day (``datetime.timedelta(1)``). You must specify a different ``schedule_interval``
+- Is your ``schedule`` param set properly? The default
```
```suggestion
- Is your ``schedule`` argument set properly? The default
```

uranusjr commented on docs/apache-airflow/faq.rst:
```diff
@@ -376,7 +376,7 @@ Why ``next_ds`` or ``prev_ds`` might not contain expected values?
 
 - When scheduling DAG, the ``next_ds`` ``next_ds_nodash`` ``prev_ds`` ``prev_ds_nodash`` are calculated using
-  ``execution_date`` and ``schedule_interval``. If you set ``schedule_interval`` as ``None`` or ``@once``,
+  ``logical_date`` and the dag's schedule interval (if applicable). If you set ``schedule`` as ``None`` or ``@once``,
```
```suggestion
  ``logical_date`` and the DAG's schedule interval (if applicable). If you set ``schedule`` as ``None`` or ``@once``,
```

uranusjr commented on docs/apache-airflow/templates-ref.rst:
```diff
@@ -89,16 +89,11 @@ existing code to use other variables instead.
-                                        ``schedule_interval`` is ``@weekly``, ``{{ next_execution_date }}``
-                                        will be ``2018-01-08 00:00:00``
+``{{ next_execution_date }}``           the logical date of the next scheduled run (if applicable)
+                                        You may be able to use ``data_interval_end`` instead.
```
```suggestion
``{{ next_execution_date }}``           the logical date of the next scheduled run (if applicable);
                                        you may be able to use ``data_interval_end`` instead
```

uranusjr commented on docs/apache-airflow/concepts/tasks.rst:
```diff
@@ -170,7 +170,7 @@ To read more about configuring the emails, see :doc:`/howto/email-config`.
 .. note::
 
-    Only scheduled tasks will be checked against SLA. For example, manually triggered tasks will not invoke an SLA miss. For more information on ``schedule_interval`` values see :doc:`DAG Run </dag-run>`.
+    Manually-triggered tasks and tasks in event-driven dags will not be checked for an SLA miss. For more information on DAG ``schedule`` values see :doc:`DAG Run </dag-run>`.
```
```suggestion
    Manually-triggered tasks and tasks in event-driven DAGs will not be checked for an SLA miss. For more information on DAG ``schedule`` values see :doc:`DAG Run </dag-run>`.
```

uranusjr committed:
Fix casing an wording

Most significantly, use "argument" instead of "param" or "parameter"
where applicable, to avoid possible confusion, since we already have a
DAG argument called "params".

uranusjr committed:
More casing and wording fixes

uranusjr commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -102,7 +101,7 @@ scheduled one interval after ``start_date``.
 .. tip::
 
-    If ``schedule_interval`` is not enough to express your DAG's schedule,
+    If a cron expression or `timedelta` object is not enough to express your DAG's schedule,
```
```suggestion
    If a cron expression or timedelta object is not enough to express your DAG's schedule,
```

uranusjr commented on docs/apache-airflow/faq.rst:
```diff
@@ -224,15 +224,15 @@ an hour after now as ``now()`` moves along.
 
 Previously, we also recommended using rounded ``start_date`` in relation to your
-``schedule_interval``. This meant an ``@hourly`` would be at ``00:00``
+DAG ``schedule``. This meant an ``@hourly`` would be at ``00:00``
```
```suggestion
DAG's ``schedule``. This meant an ``@hourly`` would be at ``00:00``
```

uranusjr commented on docs/apache-airflow/faq.rst:
```diff
@@ -376,7 +376,7 @@ Why ``next_ds`` or ``prev_ds`` might not contain expected values?
 
 - When scheduling DAG, the ``next_ds`` ``next_ds_nodash`` ``prev_ds`` ``prev_ds_nodash`` are calculated using
-  ``execution_date`` and ``schedule_interval``. If you set ``schedule_interval`` as ``None`` or ``@once``,
+  ``logical_date`` and the DAG's schedule interval (if applicable). If you set ``schedule`` as ``None`` or ``@once``,
```
```suggestion
  ``logical_date`` and the DAG's schedule (if applicable). If you set ``schedule`` as ``None`` or ``@once``,
```

uranusjr committed:
A few more formatting fixes

uranusjr committed:
Some more TaskDag conversions

uranusjr commented on airflow/models/dag.py:
```diff
@@ -313,7 +319,7 @@ class DAG(LoggingMixin):
         ``Environment`` is used to render templates as string values.
     :param tags: List of tags to help filtering DAGs in the UI.
-    :param schedule_on: List of upstream datasets if for use in triggering DAG runs.
+    :param schedule: List of upstream datasets if for use in triggering DAG runs.
```
Should be removed entirely.

uranusjr commented on airflow/models/dag.py:
```diff
@@ -236,11 +237,16 @@ class DAG(LoggingMixin):
+        accept cron string, timedelta object, Timetable, or list of Dataset objects.
+        See also :doc:`/howto/timetable`.
     :param schedule_interval: Defines how often that DAG runs, this
         timedelta object gets added to your latest task instance's
-        execution_date to figure out the next schedule
+        execution_date to figure out the next schedule.
+        Note: deprecated in Airflow 2.4; use `schedule` instead.
     :param timetable: Specify which timetable to use (in which case schedule_interval
         must not be set). See :doc:`/howto/timetable` for more information
+        Note: deprecated in Airflow 2.4; use `schedule` instead.
```
Should we just remove these entirely from the docstring? New users don’t need to know about them, and those who are using them don’t need to read the docs on them anyway.

dstandish replied:
it is tempting. _however_, i think it 's better to leave like this because a user who encounters them might look to the docstring to understand what they are, and won't be able to easily find out that they are deprecated params.  they might go on a bit of a wild goose chase to find out.

dstandish replied:
wdyt?

uranusjr replied:
One approach used elsewhere is to document the existence of deprecated arguments in the new replacement (i.e. `schedule`) and say something like _hey if you see `schedule_interval` or `timetable` they are deprecated forms and synonym of this_. This is slightly more difficult here since `DAG` has so many arguments, but would probably still be good enough.

dstandish replied:
well i'll defer to your judgment....  i.e. if you say let's remove it, i'm happy to do... just... from my perspective it seems odd.

dstandish replied:
Maybe better than what i've done currently is at least to put the deprecation warning *before* the description... anyway... lmk

uranusjr commented on airflow/models/dag.py:
```diff
@@ -460,18 +467,46 @@ def __init__(
-                raise ValueError("Param `schedule_on` must be Sequence[Dataset]")
+        self.dataset_triggers: Optional[List[Dataset]] = None
+
+        if schedule is not NOTSET:
+            if isinstance(schedule, List):
+                # if List, only support List[Dataset]
+                if any(isinstance(x, Dataset) for x in schedule):
+                    if not all(isinstance(x, Dataset) for x in schedule):
+                        raise ValueError(
+                            "If scheduling DAG with List[Dataset], all elements must be Dataset."
+                        )
+                    self.dataset_triggers = list(schedule)
+                else:
+                    raise ValueError(
+                        "Use of List object with `schedule` param is only supported for List[Dataset]."
+                    )
+            elif isinstance(schedule, Timetable):
+                timetable = schedule
+            else:  # assumed to be ScheduleIntervalArg
+                schedule_interval = schedule
```
```suggestion
        if isinstance(schedule, List):
            # if List, only support List[Dataset]
            if any(isinstance(x, Dataset) for x in schedule):
                if not all(isinstance(x, Dataset) for x in schedule):
                    raise ValueError(
                        "If scheduling DAG with List[Dataset], all elements must be Dataset."
                    )
                self.dataset_triggers = list(schedule)
            else:
                raise ValueError(
                    "Use of List object with `schedule` param is only supported for List[Dataset]."
                )
        elif isinstance(schedule, Timetable):
            timetable = schedule
        elif schedule is not None:
            schedule_interval = schedule
```

dstandish replied:
why this change @uranusjr

uranusjr replied:
Just simpler? The original implementation has some unnecessary nesting.

dstandish committed:
remove redundant param
